domains/Vault/ManageVault/Web/ViewHelpers/VaultShowViewHelper.php

```diff
@@ -38,42 +33,35 @@ public static function lastUpdatedContacts(Vault $vault): Collection
 
     public static function upcomingReminders(Vault $vault, User $user): array
     {
+        $currentDate = Carbon::now()->copy();
+        $currentDate->second = 0;
+
         // this query is a bit long and tough to do, and it could surely
         // be optimized if I knew how to properly join queries
-        // first we get all the users the vault
-        $usersInVaultIds = $vault->users->pluck('id')->toArray();
 
-        // then we get all the user notification channels for those users
-        $userNotificationChannelIds = UserNotificationChannel::whereIn('user_id', $usersInVaultIds)
-            ->select('id')
-            ->get()
-            ->unique('id')
-            ->toArray();
+        // first we get all the user notification channels for the users in the vault
+        $userNotificationChannels = $vault->users->flatMap(fn ($u) => $u->notificationChannels);
 
         // then we get all the contact reminders scheduled for those channels
-        $currentDate = Carbon::now()->copy();
-        $currentDate->second = 0;
-
-        $contactRemindersScheduled = DB::table('contact_reminder_scheduled')
-            ->whereDate('scheduled_at', '<=', $currentDate->addDays(30))
-            ->where('triggered_at', null)
-            ->whereIn('user_notification_channel_id', $userNotificationChannelIds)
-            ->get();
+        $contactRemindersScheduled = $userNotificationChannels->flatMap(
+            fn ($channel) => $channel->contactReminders()
+                ->wherePivot('scheduled_at', '<=', $currentDate->addDays(30))
+                ->wherePivot('triggered_at', null)
+                ->get()
+        );
 
         // finally, we get all the details about those reminders
         // yeah, it's painful
-        $remindersCollection = collect();
-        foreach ($contactRemindersScheduled as $contactReminderScheduled) {
-            $reminder = ContactReminder::where('id', $contactReminderScheduled->contact_reminder_id)->with('contact')->first();
+        $remindersCollection = $contactRemindersScheduled->map(function ($reminder) use ($vault, $user) {
             $contact = $reminder->contact;
 
             if ($contact->vault_id !== $vault->id) {
-                continue;
+                return null;
```

Comment:
I'm not sure this is possible, since the method returns only an array. We should change the return type to be `?array` then.

Reply:
Here we are inside the `map` function so it's ok.